Sources/IfTag.swift

```diff
@@ -100,21 +107,72 @@ final class IfExpressionParser {
   var position: Int = 0
 
-  init(components: [String], tokenParser: TokenParser) throws {
-    self.tokens = try components.map { component in
-      if let op = findOperator(name: component) {
-        switch op {
-        case .infix(let name, let bindingPower, let cls):
-          return .infix(name: name, bindingPower: bindingPower, op: cls)
-        case .prefix(let name, let bindingPower, let cls):
-          return .prefix(name: name, bindingPower: bindingPower, op: cls)
+  private init(tokens: [IfToken]) {
+    self.tokens = tokens
+  }
+  
+  static func parser(components: [String], tokenParser: TokenParser) throws -> IfExpressionParser {
+    try IfExpressionParser.validateBracketsBalance(components: components)
+    return try IfExpressionParser(components: ArraySlice(components), tokenParser: tokenParser)
+  }
+
+  static func validateBracketsBalance(components: [String]) throws {
+    guard try components.reduce(0, {
+      var bracketsBalance = $0
+      if $1 == "(" { bracketsBalance += 1 }
+      else if $1 == ")" { bracketsBalance -= 1 }
+      if bracketsBalance < 0 { throw TemplateSyntaxError("unbalanced brackets: too many closing brackets") }
+      return bracketsBalance
+    }) == 0 else {
+      throw TemplateSyntaxError("unbalanced brackets: missing closing bracket")
+    }
+  }
+  
+  private init(components: ArraySlice<String>, tokenParser: TokenParser) throws {
+    var parsedComponents = Set<Int>()
+    self.tokens = try zip(components.indices, components).flatMap { (index, component) in
+      guard !parsedComponents.contains(index) else { return nil }
+
+      if component == "(" {
+        let (expression, parsedCount) = try IfExpressionParser.subExpression(
+          from: components.suffix(from: index + 1),
+          tokenParser: tokenParser
+        )
+        parsedComponents.formUnion(Set(index...(index + parsedCount)))
+        return .subExpression(expression)
+      }
+      else if component == ")" {
+        parsedComponents.insert(index)
+        return nil
+      } else {
+        parsedComponents.insert(index)
+        if let op = findOperator(name: component) {
+          switch op {
+          case .infix(let name, let bindingPower, let cls):
+            return .infix(name: name, bindingPower: bindingPower, op: cls)
+          case .prefix(let name, let bindingPower, let cls):
+            return .prefix(name: name, bindingPower: bindingPower, op: cls)
+          }
         }
+        return .variable(try tokenParser.compileFilter(component))
       }
-
-      return .variable(try tokenParser.compileFilter(component))
     }
   }
 
+  static func subExpression(from components: ArraySlice<String>, tokenParser: TokenParser) throws -> (Expression, Int) {
+    var bracketsBalance = 1
+    let subComponents = components
+      .prefix(while: {
+        if $0 == "(" { bracketsBalance += 1 }
+        else if $0 == ")" { bracketsBalance -= 1 }
+        return bracketsBalance != 0
+      })
+    
```

Comment:
Not really sure it would be worth it, but what about a small `class Balancer` helper for every time we have to balance parentheses and/or brackets? Could serve here, in the coffee above too, and in the other PR from @djbe about indirect reference…

```swift
class Balancer {
  var balance = 0
  let opening: Character
  let closing: Character
  init(pair: (Character, Character), initialBalance: Int = 0) {
    opening = pair.0
    closing = pair.1
    balance = initialBalance
  }
  // convenience constructors
  static func brackets(initial: Int = 0) -> Balancer {
    return Balancer(pair: ("[", "]"), initialBalance: initial)
  }
  // And same for parentheses

  func parse(_ char: Character) throws -> Int {
    // … parse the char, increment/decrement the balance if it matches the opening/closing chars, throws if balance < 0, return new balance
  }
}
```
For example here you could then create a balancer then use `components.prefix(while: { balancer.parse($0) != 0)`

_(all that code typed from my phone so no guarantee of being correct or typo-free)_

I'll let you discuss if it's worth adding it? but could be nice for various places where we do that kind of logic, in this PR and others

Reply:
Good idea but I don't think it will be useful in #215 as validation there is done during parsing, not as a separate step. We can do the same here and remove validation step.